lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeNames.cpp

```diff
@@ -1499,9 +1499,42 @@ SwiftLanguageRuntime::GetGenericSignature(llvm::StringRef function_name,
   return signature;
 }
 
-std::string SwiftLanguageRuntime::GetParentNameIfClosure(StringRef name) {
+/// Returns true if a function called `symbol_name` exists in `module`.
+static bool SymbolExists(StringRef symbol_name, Module &module) {
+  SymbolContextList sc_list;
+  Module::LookupInfo lookup_info(ConstString(symbol_name),
+                                 lldb::FunctionNameType::eFunctionNameTypeFull,
+                                 lldb::eLanguageTypeSwift);
+  module.FindFunctions(lookup_info, CompilerDeclContext(),
+                       ModuleFunctionSearchOptions(), sc_list);
+  return !sc_list.IsEmpty();
+}
+
+// There are two possible variants of constructors: allocating (Kind::Allocator)
+// and initializing (Kind::Constructor). When mangling a closure, the "context"
+// is arbitrarily chosen to be the initializing variant. This function checks
+// for the existence of either in the module, starting with the Constructor
+// variant, returning whichever exists. If none exists, returns "".
+static std::string CheckIfAnyConstructorExists(StringRef ctor_kind_mangled_name,
+                                               Module &module) {
+  if (SymbolExists(ctor_kind_mangled_name, module))
+    return ctor_kind_mangled_name.str();
+  // As per swift/lib/Demangling/Demangler.cpp, Kind::Constructor is
+  // expected to end in `c` and `Kind::Allocator` to end in `C`. This is a
+  // cheap way to avoid recreating the mangled tree from scratch.
+  if (ctor_kind_mangled_name.back() == 'c') {
```

Comment:
We already do so many demangle operations when resolving types that I would rather demangle again here rather than hardcode details of the mangling scheme.

Reply:
It's not so much about demangling again, it's about reconstructing the entire demangle tree.
There is no "node.setKind", we are forced to rebuild everything.

Reply:
I think I found a way to do it.